tools/distribution/sign.py

```diff
@@ -0,0 +1,537 @@
+        response = subprocess.run(
+            self.sign_command(pkg_file), capture_output=True, encoding="utf-8")
+        self.log(response.stdout)
+
+        if response.returncode:
+            raise SigningError(response.stderr)
+
+        self.log(response.stderr)
+
+
+# Runner
+
+
+class PackageSigningRunner(runner.Runner):
+    """For a given `package_type` and `path` this will run the relevant signing
+    util for the packages they contain.
+    """
+
+    _signing_utils = ()
+
+    @classmethod
+    def register_util(cls, name: str, util: Type[DirectorySigningUtil]) -> None:
+        """Register util for signing a package type"""
+        cls._signing_utils = getattr(cls, "_signing_utils") + ((name, util),)
+
+    @property
+    def command(self) -> str:
+        """The command to sign the packages with, eg `debsign` or `rpmsign`
+
+        If specified, this will overwrite the default specified for the util,
+        and will be called with the util-specific args.
+        """
+        return self.args.command
+
+    @property
+    def extract(self) -> bool:
+        return self.args.extract
+
+    @cached_property
+    def maintainer(self) -> PackageMaintainer:
+        """A representation of the maintainer with GPG capabilities"""
+        return self.maintainer_class(self)
+
+    @property
+    def maintainer_class(self) -> Type[PackageMaintainer]:
+        return PackageMaintainer
+
+    @property
+    def maintainer_email(self) -> str:
+        """Email of the maintainer if set"""
+        return self.args.maintainer_email
+
+    @property
+    def maintainer_name(self) -> str:
+        """Name of the maintainer if set"""
+        return self.args.maintainer_name
+
+    @property
+    def package_type(self) -> str:
+        """Package type - eg deb/rpm"""
+        return self.args.package_type
+
+    @property
+    def path(self) -> str:
+        """Path to the packages directory"""
+        return self.args.path
+
+    @property
+    def tar(self) -> bool:
+        return self.args.tar
+
+    @cached_property
+    def signing_utils(self) -> dict:
+        """Configured signing utils - eg `DebSigningUtil`, `RPMSigningUtil`"""
+        return dict(getattr(self, "_signing_utils"))
+
+    def add_arguments(self, parser: argparse.ArgumentParser) -> None:
+        super().add_arguments(parser)
+        parser.add_argument(
+            "path", default="", help="Path to the directory containing packages to sign")
+        parser.add_argument(
+            "--extract",
+            action="store_true",
+            help=
+            "If set, treat the path as a tarball containing directories according to package_type")
+        parser.add_argument("--tar", help="Path to save the signed packages as tar file")
+        parser.add_argument(
+            "--type",
+            default="",
+            choices=[c for c in self.signing_utils] + [""],
+            help="Package type to sign")
+        parser.add_argument(
```

Comment:
remove this for now - it would probs work better to do something like `--type=deb:/usr/bin/mysign` so that commands are set per package type